include/bout/index_derivs_interface.hxx

```diff
@@ -308,9 +308,34 @@ template <typename T>
 T VDDY(const T& vel, const T& f, CELL_LOC outloc = CELL_DEFAULT,
        const std::string& method = "DEFAULT", const std::string& region = "RGN_NOBNDRY") {
   AUTO_TRACE();
-  const bool fHasParallelSlices = (f.hasParallelSlices());
-  const bool velHasParallelSlices = (vel.hasParallelSlices());
-  if (fHasParallelSlices && velHasParallelSlices) {
+
+  auto* localmesh = f.getMesh();
+  const CELL_LOC allowedStaggerLoc = localmesh->getAllowedStaggerLoc(DIRECTION::Y);
+  const CELL_LOC inloc = f.getLocation();
+  const CELL_LOC vloc = vel.getLocation();
+  if (outloc == CELL_DEFAULT) {
+    outloc = inloc;
+  }
+  const STAGGER stagger = localmesh->getStagger(vloc, inloc, outloc, allowedStaggerLoc);
+
+  // Some notes about the choice to be made here:
+  // - If we're not staggered, then we don't need the parallel slices
+  //   for `vel`, and we can use the `YOrthogonal` version if `f` has
+  //   parallel slices. Otherwise, we need to use the aligned version.
+  // - If we are staggered, we can only use the `YOrthogonal` version if
+  //   both `vel` and `f` have parallel slices.
+  // - If we are staggered and only `f` has parallel slices, then we
+  //   have another choice: calculate the parallel slices for `vel`
+  //   and use the `YOrthogonal` version; or throw away the slices for
+  //   `f` and use the aligned version. Which is cheaper depends on
+  //   how many parallel slices we're using. Our current estimate as
+  //   of October 2019 is if MYG > 1, it's cheaper to use the aligned
+  //   version.
```

Comment:
This is also cheaper if neither field has slices for any myg.

Reply:
I'll just mention here as well that as we only have a `{3D,3D}` and `{2D,2D}` version of this routine it's not uncommon for us to get 2D->3D implicit conversion here -- these converted fields won't have parallel slices even though they're trivial to calculate for non-FCI. This relates to issue #1811. One solution there would be to but an identity transform in the converting constructor to ensure 3D created from 2D automatically gets sensible slices, but this wouldn't work for FCI etc. 

From timing VDDY etc. for different field combinations it seems that the calculation of the derivative can actually be cheaper than copying the data from the 2D into a 3D so the most cpu efficient option here may actually be to pass in Field3D versions of your 2D data.